Реализация финального спринта №15 (add-bookings)

KanID92 opened the pull request.

KanID92 committed:
Рефакторинг по вебинару, перед следующим ТЗ:
- добавить разные типы dto;
- тип исходящего dto исправить на record;
- пробрасывание ошибок NullPointerException и проверки при обновлении перенести в слой сервиса;
- доработать логику хранения (в UserRepository - set с электронной почтой, в ItemRepository - добавить хранение по пользователям);
- убрать вызов сервиса UserService из другого сервиса (ItemService);
- добавать константу заголовка id пользователя в контроллеры.
- исправить аннотацию мапперов с UtilityClass на Component.

KanID92 committed:
Реализация финального спринта №15 (add-bookings)

KanID92 committed:
Реализация финального спринта №15 (add-bookings)

KanID92 committed:
Merge remote-tracking branch 'origin/add-bookings' into add-bookings

# Conflicts:
#	src/main/java/ru/practicum/shareit/item/repository/ItemRepository.java
#	src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java
#	src/main/java/ru/practicum/shareit/user/repository/UserRepository.java

KanID92 committed:
Merge remote-tracking branch 'origin/add-bookings' into add-bookings

# Conflicts:
#	src/main/java/ru/practicum/shareit/item/repository/ItemRepository.java
#	src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java
#	src/main/java/ru/practicum/shareit/user/repository/UserRepository.java

KanID92 committed:
Merge remote-tracking branch 'origin/add-bookings' into add-bookings

# Conflicts:
#	src/main/java/ru/practicum/shareit/item/repository/ItemRepository.java
#	src/main/java/ru/practicum/shareit/user/repository/UserRepository.java

KanID92 committed:
- корректировка по тестам GitHub (откорректировано название id моделей).

KanID92 committed:
Merge remote-tracking branch 'origin/add-bookings' into add-bookings

KanID92 committed:
Merge remote-tracking branch 'origin/add-bookings' into add-bookings

KanID92 committed:
Merge remote-tracking branch 'origin/add-bookings' into add-bookings

kesch9 commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,9 +1,80 @@
+
+    @PostMapping
+    public BookingOutputDto add(
+            @RequestBody BookingCreateDto bookingCreateDto,
```
Обязательно
Для валидации dto нужно добавить @Valid

kesch9 commented on src/main/java/ru/practicum/shareit/booking/dto/BookingCreateDto.java:
```diff
@@ -0,0 +1,23 @@
+    private Long itemId;
+
+    @NotNull
+    private LocalDateTime start;
```
Обязательно
Для валидации start нужно добавить @FutureOrPresent, а для end @Future, bookingUserId и itemId могут не прийти поэтому лучше убрать notnull, а bookingUserId вообще можно удалить т.к. это поле приходит в X-Sharer-User-Id

kesch9 commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    @Override
+    public BookingOutputDto get(long bookingId, long userId) {
```
Обязательно
Нужно добавить, что такой пользователь по userId существует перед запросом booking

kesch9 commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,146 @@
+
+    @Override
+    public BookingOutputDto get(long bookingId, long userId) {
+        Booking booking = bookingRepository.findById(bookingId)
```
Обязательно
При получении по id надо не только найти по id booking, но и отфильтровать чтобы booker и item были для этого пользователя

kesch9 commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,146 @@
+    private final ItemRepository itemRepository;
+    private final BookingRepository bookingRepository;
+
+    Sort sortByStartDesc = Sort.by(Sort.Direction.DESC, "start");
```
Отлично, что воспользовались Sort, это правильное решение

kesch9 commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,146 @@
+            case WAITING -> bookingsList = bookingRepository.findAllByBookerIdAndStatus(
+                    bookerUserId, Status.WAITING, sortByStartDesc);
+
+            case REJECTED -> bookingsList = bookingRepository.findAllByBookerIdAndStatus(
```
Обязательно
Для REJECTED еще статус CANCELED в запросе
b.status = 'CANCELED' or b.status = 'REJECTED'

kesch9 commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,146 @@
+            case ALL -> bookingsList = bookingRepository.findAllByItemOwnerIdAndStatus(
+                    ownerUserId, Status.APPROVED, sortByStartDesc);
+
+            case CURRENT -> bookingsList = bookingRepository.findAllByItemOwnerIdAndStartIsBeforeAndEndIsAfterAndStatus(
```
Обязательно
Статус не нужен при поиске не нужен для поиск, можно также переписать на Query, так более читаемо
@Query(" select b " +
			"from Booking b " +
			"where b.item.owner.id = ?1 " +
			"  and ?2 between b.start and b.end")
	List<Booking> findOwnerCurrentForDate(Long bookerId, LocalDateTime date, Sort sort);

kesch9 commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,146 @@
+        Item item = itemRepository.findById(bookingCreateDto.getItemId())
+                .orElseThrow(() -> new NotFoundException(
+                        "Item with id " + bookingCreateDto.getItemId() + " not found"));
+        if (!item.getIsAvailable()) {
```
Обязательно
Также нужно проверить доступен ли сам booking для этого нужно выполнить вот такой запрос
@Query(" select (count(b) > 0) " +
			"from Booking b " +
			"where b.item.id = ?1 " +
			"  and b.status = 'APPROVED' " +
			"  and b.start <= ?3 " +
			"  and b.end >= ?2")
	boolean isAvailableForBooking(Long itemId, LocalDateTime start, LocalDateTime end);
и по или добавить в условие

kesch9 commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -2,80 +2,154 @@
     @Override
-    public List<ItemDto> getAllUserItems(long userId) {
-        userService.getById(userId);
+    public List<ItemOutputDto> getAllUserItems(long userId) {
```
Обязательно
При поиске по userId должны заполняться комментарии и lastBooking, nextBooking
Для этого 
Выгружаем комментарии, группируем их по вещам и конвертируем в dto
Выгружаем подтвержденные бронирования отсортированные по дате начала и группируем их по вещам
Для каждой вещи находим последнее и следующее бронирование, а также комментарии и формируем результат

kesch9 replied:
Игорь вы реализовали но не совсем эффективно, надо использовать в запросе IN в котором будет itemId, так вы получите за один запрос все данные, вместо одно запроса на каждый item, таких запросов может быть очень много. Приведу пример получения комментариев
Map<Item, List<Comment>> comments = commentRepository.findByItemIn(items, Sort.by(DESC, "created"))
				.stream()
				.collect(groupingBy(Comment::getItem, toList()));
Теперь есть мапа с комментарием для item и при маппинге в dto, вы по ключе можете получить список комментариев для этого item.
Подумайте как реализовать для booking, примерно также

kesch9 commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -2,80 +2,154 @@
-                () -> new NotFoundException("Get: Item with id = " + itemId + " not found"));
-        return ItemDtoMapper.toDto(item);
+    public ItemOutputDto get(long itemId, long userId) {
+        Item item = itemRepository.findById(itemId)
```
Обязательно
В запросе должен участвовать userId
itemRepository.findByIdAndOwner_Id(itemId, userId);

kesch9 commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -2,80 +2,154 @@
+
+        List<Comment> commentList = commentRepository.findAllByItemIdOrderByCreatedAsc(itemId);
+
+        if (item.getOwnerId() != userId) {
```
Обязательно
Эта проверка не нужна, нужно написать запрос с userId

kesch9 commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -2,80 +2,154 @@
+        Item item = itemRepository.findById(itemId)
+                .orElseThrow(() -> new NotFoundException("Item with id " + itemId + " not found"));
+
+        Comment comment = CommentDtoMapper.fromCreateDto(commentCreateDto);
```
Обязательно
comment надо создавать уже после всех проверок, т.е. после проверки booking также

kesch9 commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -2,80 +2,154 @@
     @Override
-    public ItemDto add(@Valid ItemDto itemDto, long userId) {
-        userService.getById(userId);
+    public ItemOutputDto add(@Valid ItemCreateDto itemCreateDto, long userId) {
```
Обязательно
В сервисных классах нигде не проставлены аннотации @Transactional, надо проставить

kesch9 commented on src/main/java/ru/practicum/shareit/user/repository/InMemoryUserRepositoryImpl.java:
```diff
@@ -0,0 +1,62 @@
+
+@Slf4j
+@Repository
+public class InMemoryUserRepositoryImpl implements InMemoryUserRepository {
```
Можно лучше
InMemory реализации можно удалить, дальше они не понадобятся, в истории гита они останутся, но вам не придется их править если будет новые метод  и т.д.

kesch9 reviewed (commented):
Нужно еще поработать над заданием

KanID92 committed:
Корректировка по ревью (ТЗ спринт №15):
- добавить @Valid к BookingCreateDto;
- добавить аннотации валидации к BookingCreateDto;
- добавить проверку на наличие пользователя при запросе booking;
- добавить проверку booker и item в получении booking по id;
- создание Comment  методе get класса ItemServiceImpl перенесен после проверок;
- в методы поиска Item пользователю добавить заполение комментариями и крайними бронированиями;
- добавить аннотации @Transactional для методов классов сервисов;
- удалить классы реализации репозиториев в памяти.

KanID92 committed:
Корректировка по ревью (ТЗ спринт №15):
- добавить @Valid к BookingCreateDto;
- добавить аннотации валидации к BookingCreateDto;
- добавить проверку на наличие пользователя при запросе booking;
- добавить проверку booker и item в получении booking по id;
- создание Comment  методе get класса ItemServiceImpl перенесен после проверок;
- в методы поиска Item пользователю добавить заполение комментариями и крайними бронированиями;
- добавить аннотации @Transactional для методов классов сервисов;
- удалить классы реализации репозиториев в памяти.

KanID92 committed:
Merge remote-tracking branch 'origin/add-bookings' into add-bookings

# Conflicts:
#	src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java

kesch9 reviewed (approved):
Approve

KanID92 committed:
- ItemServiceImpl. Корректировка метода getAllUserItems(long userId). Уменьшить колличество запросов в БД.
- CommentRepository, BookingRepository. Добавить методы для получения списка элеметов по Item.
